src/Bicep.Core/Analyzers/Linter/Rules/OutputsShouldNotContainSecretsRule.cs

```diff
@@ -0,0 +1,142 @@
+            {
+                // Does the output name contain 'password' (suggesting it contains an actual password)?
+                if (syntax.Name.IdentifierName.ToLowerInvariant().Contains("password"))
+                {
+                    string foundMessage = string.Format(CoreResources.OutputsShouldNotContainSecretsOutputName, syntax.Name.IdentifierName);
+                    this.diagnostics.Add(parent.CreateDiagnosticForSpan(syntax.Span, foundMessage));
+                }
+
+                var visitor = new OutputValueVisitor(this.parent, diagnostics, model);
+                visitor.Visit(syntax);
+
+                // Note: No need to navigate deeper, don't call base
+            }
+        }
+
+        private class OutputValueVisitor : SyntaxVisitor
+        {
+            private List<IDiagnostic> diagnostics;
+
+            private readonly OutputsShouldNotContainSecretsRule parent;
+            private readonly SemanticModel model;
+            private const string ListFunctionPrefix = "list";
+
+            public OutputValueVisitor(OutputsShouldNotContainSecretsRule parent, List<IDiagnostic> diagnostics, SemanticModel model)
+            {
+                this.parent = parent;
+                this.model = model;
+                this.diagnostics = diagnostics;
+            }
+
+            public override void VisitVariableAccessSyntax(VariableAccessSyntax syntax)
+            {
+                // Look for references of secure parameters, e.g.:
+                //
+                //   @secure()
+                //   param secureParam string
+                //   output badResult string = 'this is the value ${secureParam}'
+
+                Symbol? symbol = model.GetSymbolInfo(syntax);
+                if (symbol is ParameterSymbol param)
+                {
+                    if (param.IsSecure())
+                    {
+                        string foundMessage = string.Format(CoreResources.OutputsShouldNotContainSecretsSecureParam, syntax.Name.IdentifierName);
+                        this.diagnostics.Add(parent.CreateDiagnosticForSpan(syntax.Name.Span, foundMessage));
+                    }
+                }
+
+                base.VisitVariableAccessSyntax(syntax);
+            }
+
+            public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
+            {
+                // Look for usage of list*() stand-alone function, e.g.:
+                //
+                //   output badResult object = listKeys(resourceId('Microsoft.Storage/storageAccounts', 'storageName'), '2021-02-01')
+                //
+
+                if (syntax.Name.IdentifierName.StartsWithOrdinalInsensitively(ListFunctionPrefix))
+                {
+                    string foundMessage = string.Format(CoreResources.OutputsShouldNotContainSecretsFunction, syntax.Name.IdentifierName);
+                    this.diagnostics.Add(parent.CreateDiagnosticForSpan(syntax.Span, foundMessage));
+                }
```

Comment:
Note that not all list* functions are necessarily returning sensitive data - see https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource#implementations.